### ISSUE TITLE
logstash-forwarder keep deleted files open

### ISSUE THREAD

**Issue by nitdana** [elastic/logstash-forwarder#200]
Hi,

When a log file is rotated the logstash-forwarder agent keeps the deleted file opened until you kill its service.
This problem obviously causes an unhealthy situation of  a consistently decreasing free space on the filesystem and currently my patch solution is scheduled restart for logstash-forwarder by a cron.
I've already seen issue #155 but unfortunately, it seems that nothing have changed since then.

Is there something else I can do to prevent this problem to occur?
Thanks!

**Comment by alexykot** [comment]
Hey, we're having issues with this thing still because it's not in the packaged version. Any idea when this could become be available in the repo?

**Comment by jordansissel** [comment]
@alexykot there is no packaged version? The repo shouldn't have logstash-forwarder at this time, if it does, please let me know where so I can fix that. The last release of lsf is old and buggy and we shouldn't be shipping it anymore.

If you need to use lsf, please build from master.

**Comment by ericcj** [comment]
we, and it sounds like others from https://github.com/elastic/logstash-forwarder/issues/155, would really appreciate having a "deleted time" option that told logstash-forwarder to give up sending to the logstash server if the file's been deleted for days.

**Comment by jordansissel** [comment]
@ericcj that's what the "dead time" setting is for.

**Comment by jordansissel** [comment]
It is my belief that the "dead time" setting solves the problem of LSF keeping files open too-long. The default dead time is 24 hours and is configurable if you need it.

If you have other issues, or continue to have issues, please open a new ticket, since I think this is fixed and there is tons of unrelated discussion in this ticket so it's hard for me to help effectively :)

**Comment by ericcj** [comment]
is there a fix past version 0.4 that solves it?  when i posted that comment i was already using the default dead time setting from version 0.4 and it was definitely keeping deleted files open.  wouldn't it just hang in readline forever on rotated files and never close them since nothing is ever written to them again to make progress to the line that's supposed to close?

**Comment by jordansissel** [comment]
@ericcj 0.4.0 (and some older I think) includes the feature ("dead time") that solves this. If you observe this _not_ working, then that's a new bug, and not this issue, and is worth filing a new ticket for - does that make sense?

**Comment by abordia** [comment]
How do i handle above issue for a logstash (not logstash forwarder)?  I am unable to specify "dead time" in logstash 1.4.2.  I see logstash process holding 'open files' for extended period of time.  It seemed forever since logstash never recovered over a period of days.

I see many (say 10) similar entries:
java      1802    root   64r      REG              202,2    190416     150315 /var/log/messages-20150891 (deleted) 
This does not recover in 24 hours.  Files it is pointing to does not exist (they are gzipped).

**Comment by abordia** [comment]
@jordansissel How do i handle above issue for a logstash (not logstash forwarder)? I am unable to specify "dead time" in logstash 1.4.2. I see logstash process holding 'open files' for extended period of time. It seemed forever since logstash never recovered over a period of days.

I see many (say 10) similar entries:
java 1802 root 64r REG 202,2 190416 150315 /var/log/messages-20150891 (deleted) 
This does not recover in 24 hours. Files it is pointing to does not exist (they are gzipped).

**Comment by driskell** [comment]
@abordia One thing to consider is that there is another prerequisite for closing a file and that is that all logs are ingested/shipped. It could be your filters or outputs just can't keep up.

**Comment by abordia** [comment]
@driskell My understanding: Logstash is still reading while file gets zipped up and deleted and hence it keeps the file handler open?  Eventually logstash stops sending any messages to ES.
How do i recover w/o restarting logstash?

**Comment by abordia** [comment]
@driskell My logstash cpu is still churning (7+ days) and not sending anything to ES (no network traffic detected).   I have detected file handlers pointing to 'deleted' files as stated earlier (from 2 -15 oopen pointers to deleted file). Underlying files were gzipped.
My Q: 1.  How do i reduce the "dead time" from 24 hours in LS 1.4
           2.  If issue is that "LS has not finished ingesting/ shipping" logs.  How can this be fixed?  Is there a timeout that kicks in?  On few of my servers, LS cpu spinning is going on for many days. Thanks!

**Comment by luhkevin** [comment]
Is this "dead time" parameter documented anywhere? I could not find it in the README.

**Comment by ruflin** [comment]
@luhkevin dead time in filebeat is now ignore_older. A description can be found here: https://www.elastic.co/guide/en/beats/filebeat/current/filebeat-configuration-details.html#_ignore_older

**Comment by rahulghanate** [comment]
Facing the same issue,
# lsof | grep logstash | grep deleted | wc -l
XXX

Using LSF 0.4.0

**Comment by tsg** [comment]
@rahulghanate You should consider switching to [Filebeat](https://www.elastic.co/downloads/beats/filebeat). We've been focusing our development efforts on it for the last ~9 months.

**Comment by rahulghanate** [comment]
Thanks for reply.
Will think about it. Wanted to save efforts of migration if this issue would have resolved.

**Comment by fxtentacle** [comment]
After investigating this issue happening on an Ubuntu 16.04.4 LTS, I can confirm that the fix in LSF 0.4.0 is working correctly. `reader.ReadBytes` will return `io.EOF` once it has reached the end of the file, which in turn causes the `Harvester.readline` to time out after 10 seconds. At that point, the "dead time" timeout is checked and if it has been exceeded, the deleted file is closed. If not, the read is retried after one second of wait time.

So then how can you still see the files remaining open? That is due to the backpressure implemented in the logstash network protocol. If the logstash server receiving your events cannot index them quickly enough, then logstash-forwarder will delay sending new events. That, in turn, means that logstash-forwarder will delay reading new events, which in turn delays reaching EOF for deleted files.

**To summarize**, the "dead time" option will only delete a file if it has been idle for that time duration AFTER the logstash server has indexed the last event from said file. So if your logstash server is too slow with indexing, then logstash-forwarder will wait and thus keep deleted files open for much longer than the duration specified with the "dead time" option.
